daemon/migration.go

```diff
@@ -0,0 +1,143 @@
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/log"
+	"github.com/containerd/platforms"
+	"github.com/docker/docker/api/types/backend"
+	"github.com/docker/docker/container"
+	"github.com/docker/docker/image"
+	"github.com/docker/docker/internal/multierror"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+func migrateContainerOS(ctx context.Context,
+	migration platformReader,
+	ctr *container.Container,
+) {
+	deduced, err := deduceContainerPlatform(ctx, migration, ctr)
+	if err != nil {
+		log.G(ctx).WithFields(log.Fields{
+			"container": ctr.ID,
+			"error":     err,
+		}).Warn("failed to deduce the container architecture")
+		ctr.ImagePlatform.OS = ctr.OS //nolint:staticcheck // ignore SA1019
+		return
+	}
+
+	ctr.ImagePlatform = deduced
+}
+
+type platformReader interface {
+	ReadPlatformFromConfigByImageManifest(ctx context.Context, desc ocispec.Descriptor) (ocispec.Platform, error)
+	ReadPlatformFromImage(ctx context.Context, id image.ID) (ocispec.Platform, error)
+}
+
+// deduceContainerPlatform tries to deduce `ctr`'s platform.
+// If both `ctr.OS` and `ctr.ImageManifest` are empty, it's not possible to deduce
+// the container platform and an error is returned.
```

Comment:
Not sure if we still need to bother with that (probably-not), but ISTR that because pre-historic images didn't have platform (there was only Linux, and there was only x86), some codepaths treated "if nothing present, it must be `linux/amd64`).

Reply:
The pre-historic image case is already handled by `container.FromDisk` directly where it just defaults to host platform (previous behavior was to default to host OS).

In practice it shouldn't really matter as this branch of `deduceContainerPlatform` shouldn't be reached at all.
I'll adjust the comment though as it's wrong 😅  In this case it explicitly returns a default platform to match the pre-historic image behavior.